Imitation learning

alexander-zap opened the pull request:
Support to use imitation learning algorithms (and pass recorded expert sequences to learn from) for agent training.

alexander-zap committed:
Added support for imitation learning

- split of project structure (split of RL and IL agent class)
- .train signature change: IL algorithms require passing recorded episode sequences to learn from, while RL algorithms require passing of the environment itself
- implemented ImitationAgent which wraps BC algorithm from imitation library

alexander-zap committed:
Uploaded example script for imitation learning

alexander-zap committed:
Made training_timesteps a mandatory parameter again

alexander-zap committed:
BC using timesteps

Luke100000 commented on exploration/train_il_agent.py:
```diff
@@ -0,0 +1,101 @@
+    )
+
+    episode_sequences = []
+    for trajectory in rollouts:
```
This conversion could be the job of the discussed trajectory dataset wrapper/container (e.g., from_buffer(rollouts).iterate_tupple_trajectories())

alexander-zap replied:
Yes, this will be implemented as a functionality in the dataset class.

alexander-zap committed:
hotfix: HF connector not reporting mean_reward or std_reward if they are 0.0

alexander-zap committed:
bump poetry.lock

alexander-zap committed:
Major refactoring and addition to imitation framework intergration

- support for all available DemonstrationAlgorithms through AlgorithmWrapper class
- introduced algorithm_policy attribute to store policies independent of (non-initialized) algorithms; important for saving and loading, since algorithm initialization requires calling the train function
- added vectorization of environment into train function (required for most algorithms)
- added saving and loading function which work by zipping/unzipping all required individual components of the imitation algorithm

alexander-zap committed:
Adjust IL exploration script to test all features (including the creation of a imitation dataset)

alexander-zap committed:
remove auto-deletion of upload files from connector (leads to conflict with temporary files)

alexander-zap committed:
minor documentation fix

alexander-zap committed:
minor improvement of exploration script (usage of gym.make, correct setting of download boolean)

alexander-zap committed:
Fixed SQIL loading not working (requires re-initialization of RL algo with environment and buffer)

alexander-zap committed:
Fixed correct passing of algorithm_params at loading time

alexander-zap committed:
Added default training steps to imitation exploration script

alexander-zap committed:
Bug-fix: Use all provided algorithm_parameters

alexander-zap committed:
Add "policy" to SB3 parameters in any way (if not provided by user)

alexander-zap committed:
Allow using variable horizon lengths (which are present in most environments)

alexander-zap committed:
Switched environment to regular one (no fixed time horizon)

alexander-zap committed:
Raise error when saving an agent without the algorithm being initialized (also better logging for "bad loading").

alexander-zap committed:
load expert for CartPole-v1 in imitation exploration script

alexander-zap committed:
remove "allow_variable_horizon" from parameters for BC and SQIL (and print warning instead)

Luke100000 reviewed (approved):
LGTM
